### PR TITLE
FIPS tests added for gnutls library

### PR DESCRIPTION
**Pull request by rcmadhankumar**
[CI:TOXENVS] fips

fixes: #495

### CODE REVIEW

File: tests/test_fips.py [hunk 1]
```diff
@@ -152,3 +187,52 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
...
+        "sha224": "54655eae3d97147de34564572231c34d6d0917dd7852b5b93647fb4fe53ee97e5e0a2a4d359b5b461409dc44d9315afbc3b7d6bc5cd598e6",
+        "sha256": "4ea6a95a3a56fa6b7c1673c145198c52265fea4fe4cebef97249b39c25a733a0d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+        "sha384": "416b69644f4844065fcfe7b60f14fe07d1573420c4db2b15a6d1f4cb2b6933ffce8dcb1bce7788b962eb3d0c8d4eefc4acbfd470c22c0d95a1d10a087dc31988b9f7bfeb13be70b876a73558be664e5858d11f9459923e6e5fd838cb5708b969",
+        "sha512": "3916af571551c0c40eb19936c7ac2c090e140cad48e348dc4e9d5b6508a34ac6090daeeed3a081ce0e44c90b181987b71f09e8e0c190e5af26cd46eea724489de1c112ff908febc3b98b1693a6cd3564eaf8e5e6ca629d084d9f0eba99247cacdd72e369ff8941397c2807409ff66be64be908da17ad7b8a49a2a26c0e8086aa",
+    }
+
+    for digest in FIPS_GNUTLS_DIGESTS:
+        container_per_test.connection.check_output(
+            f"/bin/fips-test-gnutls {digest}"
+        )
```
Comment (dcermak) [inline tests/test_fips.py:225]: No need to call `check_output` twice
```suggestion
```

File: tests/files/fips-test-gnutls.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    if (gnutls_hash_fast(digest_algorithm, (const unsigned char *)mess2, strlen(mess2), hash2) != GNUTLS_E_SUCCESS) {
+        fprintf(stderr, "Hash calculation failed\n");
+        goto cleanup;
+    }
+
+    memcpy(md_value, hash1, hash1_size);
+    memcpy(md_value + hash1_size, hash2, hash2_size);
+    md_len = hash1_size + hash2_size;
+
+    gnutls_global_deinit();
```
Comment (dcermak) [inline tests/files/fips-test-gnutls.c:53]: Your program will now always reach `cleanup`, which already calls `gnutls_global_deinit`
```suggestion
```

File: tests/test_fips.py (hunk 1 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: fixed.

File: tests/files/fips-test-gnutls.c (hunk 2 above)
Comment (rcmadhankumar) [reply tests/files/fips-test-gnutls.c]: fixed.

File: tests/test_fips.py [hunk 3]
```diff
@@ -152,3 +187,49 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
 )
 def test_openssl_fips_hashes(container_per_test: ContainerData):
     openssl_fips_hashes_test_fnct(container_per_test)
+
+
+@pytest.mark.parametrize(
+    "container_per_test", FIPS_GNUTLS_TESTER_IMAGES, indirect=True
+)
+def test_gnutls_binary(container_per_test: ContainerData) -> None:
+    """Check that a binary linked against Gnutls obeys the host's FIPS mode
```
Comment (alexandrevicenzi) [inline tests/test_fips.py:196]: nit: GnuTLS not Gnutls

File: tests/test_fips.py [hunk 4]
```diff
@@ -152,3 +187,49 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
...
+        "sha256": "4ea6a95a3a56fa6b7c1673c145198c52265fea4fe4cebef97249b39c25a733a0d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+        "sha384": "416b69644f4844065fcfe7b60f14fe07d1573420c4db2b15a6d1f4cb2b6933ffce8dcb1bce7788b962eb3d0c8d4eefc4acbfd470c22c0d95a1d10a087dc31988b9f7bfeb13be70b876a73558be664e5858d11f9459923e6e5fd838cb5708b969",
+        "sha512": "3916af571551c0c40eb19936c7ac2c090e140cad48e348dc4e9d5b6508a34ac6090daeeed3a081ce0e44c90b181987b71f09e8e0c190e5af26cd46eea724489de1c112ff908febc3b98b1693a6cd3564eaf8e5e6ca629d084d9f0eba99247cacdd72e369ff8941397c2807409ff66be64be908da17ad7b8a49a2a26c0e8086aa",
+    }
+
+    for digest in FIPS_GNUTLS_DIGESTS:
+        res = container_per_test.connection.check_output(
+            f"/bin/fips-test-gnutls {digest}"
+        )
+        assert expected_fips_gnutls_digests[digest] in res
```
Comment (alexandrevicenzi) [inline tests/test_fips.py:226]: nit: I would prefer an exact message output match.

Something like:

`assert res == "Digest is: %s".format(expected_fips_gnutls_digests[digest])`

File: tests/test_fips.py [hunk 5]
```diff
@@ -28,6 +30,11 @@
 COPY tests/files/fips-test.c /src/
 """
 
+DOCKERFILE_GNUTLS = """WORKDIR /src/
+COPY tests/files/fips-test-gnutls.c /src/
+ENV GNUTLS_FORCE_FIPS_MODE=1
```
Comment (dcermak) [inline tests/test_fips.py:35]: This must come from the base image
```suggestion
```

**Review by dcermak — CHANGES_REQUESTED** [review]
We must not set the environment variable `GNUTLS_FORCE_FIPS_MODE` ourselves, this should come from the base FIPS image. Please xfail the failing cases instead

File: bci_tester/fips.py [hunk 6]
```diff
@@ -64,6 +64,28 @@
...
+    "sha224",
+    "sha256",
+    "sha384",
+    "sha512",
+)
+
+#: all digests supported by gnutls
+ALL_GNUTLS_DIGESTS: Tuple[str] = NONFIPS_GNUTLS_DIGESTS + FIPS_GNUTLS_DIGESTS
+
+assert len(set(ALL_GNUTLS_DIGESTS)) == len(ALL_GNUTLS_DIGESTS)
```
[the remark this answers is not in the document]
Comment (rcmadhankumar) [reply bci_tester/fips.py]: fixed.

File: tests/files/fips-test-gnutls.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+int main(int argc, char *argv[]) {
+    gnutls_digest_algorithm_t digest_algorithm;
+    const char *mess1 = "Test Message\n";
+    const char *mess2 = "Hello World\n";
+    unsigned char md_value[128];
+    size_t md_len;
+    int ret_code=1;
+
+    if (argc < 2) {
+        fprintf(stderr, "Usage: mdtest digestname\n");
```
[the remark this answers is not in the document]
Comment (rcmadhankumar) [reply tests/files/fips-test-gnutls.c]: fixed,
info was gathered from multiple sources, this particular section was copied from existing  tests/files/fips-test.c.
(Usage is wrong there, just saw it now. they might have done it as output binary name might change during compilation)

File: tests/files/fips-test-gnutls.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    memcpy(md_value + hash1_size, hash2, hash2_size);
+    md_len = hash1_size + hash2_size;
+
+    printf("Digest is: ");
+    for (size_t i = 0; i < md_len; i++) {
+        printf("%02x", md_value[i]);
+    }
+    printf("\n");
+    ret_code=0;
+    goto cleanup;
```
[the remark this answers is not in the document]
Comment (rcmadhankumar) [reply tests/files/fips-test-gnutls.c]: fixed.

File: tests/test_fips.py (hunk 3 above)
Comment (rcmadhankumar) [reply tests/test_fips.py]: fixed.

**Review by alexandrevicenzi — DISMISSED** [review]

File: tests/test_fips.py [hunk 9]
```diff
@@ -152,3 +186,53 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
 )
 def test_openssl_fips_hashes(container_per_test: ContainerData):
     openssl_fips_hashes_test_fnct(container_per_test)
+
+
+@pytest.mark.parametrize(
+    "container_per_test", FIPS_GNUTLS_TESTER_IMAGES, indirect=True
+)
+@pytest.mark.skipif(
```
Comment (dcermak) [inline tests/test_fips.py:194]: We don't want to skip this unconditionally, the image should be fixed eventually:

```suggestion
@pytest.mark.xfail(
```

File: tests/test_fips.py [hunk 10]
```diff
@@ -152,3 +186,53 @@ def fips_mode_setup_check(container_per_test: ContainerData) -> None:
...
+        "sha256": "4ea6a95a3a56fa6b7c1673c145198c52265fea4fe4cebef97249b39c25a733a0d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+        "sha384": "416b69644f4844065fcfe7b60f14fe07d1573420c4db2b15a6d1f4cb2b6933ffce8dcb1bce7788b962eb3d0c8d4eefc4acbfd470c22c0d95a1d10a087dc31988b9f7bfeb13be70b876a73558be664e5858d11f9459923e6e5fd838cb5708b969",
+        "sha512": "3916af571551c0c40eb19936c7ac2c090e140cad48e348dc4e9d5b6508a34ac6090daeeed3a081ce0e44c90b181987b71f09e8e0c190e5af26cd46eea724489de1c112ff908febc3b98b1693a6cd3564eaf8e5e6ca629d084d9f0eba99247cacdd72e369ff8941397c2807409ff66be64be908da17ad7b8a49a2a26c0e8086aa",
+    }
+
+    for digest in FIPS_GNUTLS_DIGESTS:
+        res = container_per_test.connection.check_output(
+            f"/bin/fips-test-gnutls {digest}"
+        )
+        assert "Digest is: " + expected_fips_gnutls_digests.get(digest) in res
```
Comment (dcermak) [inline tests/test_fips.py:229]: I'd recommend to not use `get()` as that always returns something (`None` if the key is not in the dictionary), which could lead to harder to understand errors
```suggestion
        assert "Digest is: " + expected_fips_gnutls_digests[digest] in res
```

Comment (rcmadhankumar) [reply tests/test_fips.py]: fixed.

**Review by dcermak — APPROVED** [review]

**Pull request merged**